Backported some changes from GMP 6.3.0, since modern GCC is much more strict
by default and doesn't like this K&R-style stuff.

New GMP has further updates to these tests (e.g., checks for __cplusplus), but
I don't think those are necessary for our purposes.

-- throwaway96, 2025-09-01

--- gmp-4.2.1.orig/acinclude.m4	2006-05-04 12:39:01.000000000 -0400
+++ gmp-4.2.1/acinclude.m4	2025-09-01 02:23:40.613367719 -0400
@@ -516,6 +516,7 @@
    values being spilled into floating point registers.  The problem doesn't
    show up all the time, but has occurred enough in GMP for us to reject
    this compiler+flags.  */
+#include <string.h>  /* for memcpy */
 struct try_t
 {
  char dst[2];
@@ -592,11 +593,12 @@
 
 #ifdef __GNUC__
 typedef unsigned long long t1;typedef t1*t2;
+void g(int a,t1 const*b,t1 c,t2 d,t1 const*e,int f){}
+void h(){}
 __inline__ t1 e(t2 rp,t2 up,int n,t1 v0)
 {t1 c,x,r;int i;if(v0){c=1;for(i=1;i<n;i++){x=up[i];r=x+1;rp[i]=r;}}return c;}
-f(){static const struct{t1 n;t1 src[9];t1 want[9];}d[]={{1,{0},{1}},};t1 got[9];int i;
+void f(){static const struct{t1 n;t1 src[9];t1 want[9];}d[]={{1,{0},{1}},};t1 got[9];int i;
 for(i=0;i<1;i++){if(e(got,got,9,d[i].n)==0)h();g(i,d[i].src,d[i].n,got,d[i].want,9);if(d[i].n)h();}}
-h(){}g(){}
 #else
 int dummy;
 #endif
@@ -608,8 +610,9 @@
    1666 to get an ICE with -O1 -mpowerpc64.  */
 
 #ifdef __GNUC__
-f(int u){int i;long long x;x=u?~0:0;if(x)for(i=0;i<9;i++);x&=g();if(x)g();}
-g(){}
+int g();
+void f(int u){int i;long long x;x=u?~0:0;if(x)for(i=0;i<9;i++);x&=g();if(x)g();}
+int g(){return 0;}
 #else
 int dummy;
 #endif
@@ -3659,7 +3662,7 @@
 int
 main ()
 {
-  exit(0);
+  return 0;
 }
 EOF
 gmp_compile="$1 conftest.c"
@@ -3733,7 +3736,7 @@
 int
 main ()
 {
-  exit (0);
+  return 0;
 }
 EOF
 for i in .exe ,ff8 ""; do
@@ -3767,7 +3770,7 @@
 int
 main (int argc, char *argv[])
 {
-  exit(0);
+  return 0;
 }
 EOF
 gmp_compile="$CC_FOR_BUILD conftest.c"
@@ -3798,10 +3801,11 @@
 AC_CACHE_CHECK([for build system compiler math library],
                gmp_cv_check_libm_for_build,
 [cat >conftest.c <<EOF
+#include <math.h>
 int
 main ()
 {
-  exit(0);
+  return 0;
 }
 double d;
 double
